cumulusci/tasks/bulkdata/data_generation/data_generator_runtime.py

```diff
@@ -0,0 +1,236 @@
+    def generate_id(self):
+        self.current_id = self.globals.id_manager.generate_id(self.current_table_name)
+        return self.current_id
+
+    def register_object(self, obj, name=None):
+        self.obj = obj
+        self.globals.register_object(obj, name)
+
+    def register_intertable_reference(self, table_name_from, table_name_to, fieldname):
+        self.globals.register_intertable_reference(
+            table_name_from, table_name_to, fieldname
+        )
+
+    def register_field(self, field_name, field_value):
+        """Register each field value to be ready to inject them into template"""
+        self.field_values[field_name] = field_value
+
+    def field_vars(self):
+        return {
+            "id": self.current_id,
+            "this": self.obj,
+            "today": self.today,
+            "fake": faker_template_library,
+            "fake_i18n": lambda locale: FakerTemplateLibrary(locale),
+            "number": self.counter,
+            **self.options,
+            **self.globals.object_names,
+            **self.field_values,
+        }
+
+    def field_funcs(self):
+        def curry(func):
+            rc = partial(func, self)
+            if hasattr(func, "lazy"):
+                rc.lazy = func.lazy
+            return rc
+
+        funcs = {name: curry(func) for name, func in template_funcs.items()}
+        return {**funcs}
+
+    def executable_blocks(self):
+        return {**self.field_funcs(), "fake": self.fake}
+
+    def fake(self, name):
+        return str(getattr(faker_template_library, name))
+
+    def get_evaluator(self, definition: str):
+        return self.template_evaluator_factory.get_evaluator(definition)
+
+
+class DynamicEvaluator:
+    def __init__(self, template):
+        self.template = template
+
+    def __call__(self, context):
+        return self.template.render(**context.field_vars(), **context.field_funcs())
+
+
+def evaluate_function(func, args: Sequence, kwargs: Mapping, context):
+    if not hasattr(func, "lazy"):
+        args = [arg.render(context) if hasattr(arg, "render") else arg for arg in args]
+        kwargs = {
+            name: arg.render(context) if hasattr(arg, "render") else arg
+            for name, arg in kwargs.items()
+        }
+    value = func(*args, **kwargs)
+    return value
+
+
+def fix_exception(message, parentobj, e):
+    """Add filename and linenumber to an exception if needed"""
+    filename, line_num = parentobj.filename, parentobj.line_num
+    if isinstance(e, DataGenError):
+        if not e.filename:
+            e.filename = filename
+        if not e.line_num:
+            e.line_num = line_num
+        raise e
+    else:
+        raise DataGenError(message, filename, line_num) from e
+
+
+class ObjectRow:
+    """Represents a single row
+
+    Uses __getattr__ so that the template evaluator can use dot-notation."""
+
+    def __init__(self, tablename, values=()):
+        self._tablename = tablename
+        self._values = values
```

Comment:
Would it be worth using `__slots__` to avoid the memory overhead of an instance dict?